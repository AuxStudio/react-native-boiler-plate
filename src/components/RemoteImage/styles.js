import { StyleSheet } from 'react-native';

import styleConstants from '../../styleConstants';

const styles = StyleSheet.create({
  container: {},
  image: {},
  backgroundContainer: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: styleConstants.colors.dividerColor,
  },
  icon: {
    fontSize: styleConstants.fonts.sizes.icon,
    color: styleConstants.colors.primaryText,
  },
});

export default styles;
