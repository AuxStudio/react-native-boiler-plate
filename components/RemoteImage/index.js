import React from 'react';
import PropTypes from 'prop-types';
import { View, ViewPropTypes, Text, ActivityIndicator } from 'react-native';
import FastImage from 'react-native-fast-image';
import Icon from 'react-native-vector-icons/MaterialIcons';

import styles from './styles';
import styleConstants from '../../styleConstants';

export default class RemoteImage extends React.Component {
  constructor(props) {
    super(props);

    this.setError = this.setError.bind(this);
    this.setLoading = this.setLoading.bind(this);

    this.state = {
      isLoading: true,
      hasError: false,
    };
  }

  static propTypes = {
    source: PropTypes.shape({
      uri: PropTypes.string,
    }),
    borderRadius: PropTypes.number,
    iconStyle: Text.propTypes.style,
    style: ViewPropTypes.style,
    loaderColor: PropTypes.string,
  };

  static defaultProps = {
    loaderColor: styleConstants.colors.primaryText,
  };

  setError() {
    this.setState({
      hasError: true,
    });

    this.setLoading(false);
  }

  setLoading(isLoading) {
    this.setState({
      isLoading,
    });
  }

  render() {
    const { hasError, isLoading } = this.state;
    const { borderRadius, iconStyle, loaderColor, source, style } = this.props;
    const borderRadiusStyles = borderRadius && { borderRadius };
    let backgroundComponent;

    if (hasError) {
      backgroundComponent = (
        <View style={[styles.backgroundContainer, borderRadiusStyles]}>
          <Icon name="error-outline" style={[styles.icon, iconStyle]} />
        </View>
      );
    } else if (isLoading) {
      backgroundComponent = (
        <View style={[styles.backgroundContainer, borderRadiusStyles]}>
          <ActivityIndicator size="large" color={loaderColor} />
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <FastImage
          source={source}
          style={[styles.image, borderRadiusStyles, style]}
          onLoadEnd={() => this.setLoading(false)}
          onError={this.setError}
        />
        {backgroundComponent}
      </View>
    );
  }
}
