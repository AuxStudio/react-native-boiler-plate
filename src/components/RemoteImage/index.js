import React from 'react';
import PropTypes from 'prop-types';
import { View, ViewPropTypes, Text, ActivityIndicator } from 'react-native';
import FastImage from 'react-native-fast-image';
import Icon from 'react-native-vector-icons/MaterialIcons';

import styles from './styles';
import styleConstants from '../../styleConstants';

export default class RemoteImage extends React.Component {
  constructor(props) {
    super(props);

    this.setHasError = this.setHasError.bind(this);
    this.setIsLoading = this.setIsLoading.bind(this);

    this.state = {
      isLoading: true,
      hasError: false,
    };
  }

  static propTypes = {
    source: PropTypes.shape({
      uri: PropTypes.string,
    }),
    borderRadius: PropTypes.number,
    iconStyle: Text.propTypes.style,
    style: ViewPropTypes.style,
    loaderColor: PropTypes.string,
  };

  static defaultProps = {
    loaderColor: styleConstants.colors.primaryText,
  };

  setHasError() {
    this.setState({
      hasError: true,
    });

    this.setIsLoading(false);
  }

  setIsLoading(isLoading) {
    this.setState({
      isLoading,
    });
  }

  render() {
    const { hasError, isLoading } = this.state;
    const { borderRadius, iconStyle, loaderColor, source, style } = this.props;
    const borderRadiusStyles = borderRadius && { borderRadius };
    let backgroundComponent;

    if (hasError) {
      backgroundComponent = (
        <View style={[styles.backgroundContainer, borderRadiusStyles]}>
          <Icon name="error-outline" style={[styles.icon, iconStyle]} />
        </View>
      );
    } else if (isLoading) {
      backgroundComponent = (
        <View style={[styles.backgroundContainer, borderRadiusStyles]}>
          <ActivityIndicator size="large" color={loaderColor} />
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <FastImage
          source={source}
          style={[styles.image, borderRadiusStyles, style]}
          onLoadEnd={() => this.setIsLoading(false)}
          onError={this.setHasError}
        />
        {backgroundComponent}
      </View>
    );
  }
}
