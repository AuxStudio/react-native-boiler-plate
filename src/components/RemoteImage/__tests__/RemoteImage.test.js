import React from 'react';
import renderer from 'react-test-renderer';

import RemoteImage from '..';

describe('RemoteImage', () => {
  const spies = [];
  const source = { uri: 'https://homepages.cae.wisc.edu/~ece533/images/airplane.png' };

  describe('renders', () => {
    it('renders with minimum required props', () => {
      const component = renderer.create(<RemoteImage />);

      expect(component).toMatchSnapshot();
    });

    it('renders the normal state', () => {
      const component = renderer.create(<RemoteImage source={source} />);

      expect(component).toMatchSnapshot();
    });

    it('renders the loading state', () => {});

    it('renders the error state', () => {});
  });

  describe('methods', () => {
    it('should handle setHasError', () => {
      const component = renderer.create(<RemoteImage />);
      const instance = component.getInstance();

      instance.setHasError(true);

      expect(instance.state.hasError).toEqual(true);
    });

    it('should handle setIsLoading', () => {
      const component = renderer.create(<RemoteImage />);
      const instance = component.getInstance();

      instance.setIsLoading(true);

      expect(instance.state.isLoading).toEqual(true);
    });
  });

  afterEach(() => {
    spies.forEach((spy) => {
      if (spy) {
        spy.mockClear();
      }
    });
  });
});
