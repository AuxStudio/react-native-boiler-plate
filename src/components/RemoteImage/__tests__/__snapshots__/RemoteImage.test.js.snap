// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteImage renders renders the normal state 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Array [
          Object {},
          undefined,
          undefined,
        ],
      ]
    }
  >
    <FastImageView
      onFastImageError={[Function]}
      onFastImageLoadEnd={[Function]}
      resizeMode="cover"
      source={
        Object {
          "uri": "https://homepages.cae.wisc.edu/~ece533/images/airplane.png",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#E0E0E0",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="#212121"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;

exports[`RemoteImage renders renders with minimum required props 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Array [
          Object {},
          undefined,
          undefined,
        ],
      ]
    }
  >
    <FastImageView
      onFastImageError={[Function]}
      onFastImageLoadEnd={[Function]}
      resizeMode="cover"
      source={null}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#E0E0E0",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="#212121"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;
